pkg/backup/backup.go

```diff
@@ -174,94 +175,137 @@ func (b *Backup) Run() {
+	bs := backupapi.BackupStatus{
+		CreationTime:     time.Now().Format(time.RFC3339),
+		Size:             util.ToMB(n),
+		Version:          etcdVersion,
+		Revision:         rev,
+		TimeTookInSecond: int(time.Since(start).Seconds() + 1),
+	}
+	b.recentBackupsStatus = append(b.recentBackupsStatus, bs)
+	if len(b.recentBackupsStatus) > maxRecentBackupStatusCount {
+		b.recentBackupsStatus = b.recentBackupsStatus[1:]
 	}
 
-	log.Printf("saving backup for cluster (%s)", b.clusterName)
-	if err := b.writeSnap(member, rev); err != nil {
-		err = fmt.Errorf("write snapshot failed: %v", err)
-		return lastSnapRev, err
+	return 0, nil
+}
+
+// WriteSnap writes the snapshot to the backend.
+func WriteSnap(version string, rev int64, be backend.Backend, rc io.ReadCloser) (int64, error) {
+	n, err := be.Save(version, rev, rc)
+	if err != nil {
+		return 0, err
 	}
-	return rev, nil
+	return n, nil
 }
 
-func (b *Backup) writeSnap(m *etcdutil.Member, rev int64) error {
-	start := time.Now()
+// GetSnap gets the snapshot from the cluster member with the maximum revision.
+func GetSnap(ctx context.Context, kclient kubernetes.Interface, tls *tls.Config, namespace, clusterName string) (io.ReadCloser, string, int64, error) {
+	pods, err := getRunningPods(kclient, namespace, clusterName)
+	if err != nil {
+		return nil, "", 0, fmt.Errorf("failed to retrieve running etcd pods: (%v)", err)
+	}
+	if len(pods) == 0 {
+		return nil, "", 0, fmt.Errorf("no running etcd pods found")
+	}
+
+	m, rev := getMemberWithMaxRev(pods, tls)
+	if m == nil {
+		return nil, "", 0, fmt.Errorf("no reachable member")
+	}
 
 	cfg := clientv3.Config{
 		Endpoints:   []string{m.ClientURL()},
 		DialTimeout: constants.DefaultDialTimeout,
-		TLS:         b.etcdTLSConfig,
+		TLS:         tls,
 	}
 	etcdcli, err := clientv3.New(cfg)
 	if err != nil {
-		return fmt.Errorf("failed to create etcd client (%v)", err)
+		return nil, "", 0, fmt.Errorf("failed to create etcd client (%v)", err)
 	}
-	defer etcdcli.Close()
-
-	ctx, cancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
-	resp, err := etcdcli.Maintenance.Status(ctx, m.ClientURL())
-	cancel()
+	// when GetSnap fails, close etcdcli.
+	defer func() {
+		if err != nil {
+			etcdcli.Close()
+		}
+	}()
+	cctx, ccancel := context.WithTimeout(context.Background(), constants.DefaultRequestTimeout)
+	resp, err := etcdcli.Maintenance.Status(cctx, m.ClientURL())
+	ccancel()
 	if err != nil {
-		return err
+		return nil, "", 0, fmt.Errorf("failed to get member status (%v)", err)
 	}
 
-	ctx, cancel = context.WithTimeout(context.Background(), constants.DefaultSnapshotTimeout)
 	rc, err := etcdcli.Maintenance.Snapshot(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to receive snapshot (%v)", err)
+		return nil, "", 0, fmt.Errorf("failed to get member snapshot (%v)", err)
 	}
-	defer cancel()
-	defer rc.Close()
+	// when context for getting/saving snapshot finishes, close the etcdcli.
+	go func() {
```

Comment:
remove this. close etcdcli if it's not needed anymore

Reply:
we can't close `etcdcli` if `WriteSnap` is not done with `rc`. Closing  `etcdcli` before `WriteSnap` results rc being closed.